File: android/src/main/java/xyz/justzappit/zappmessaging/ZMLog.kt
package xyz.justzappit.zappmessaging

import android.util.Log

/**
 * Privacy-safe native diagnostics.
 *
 * Native logging follows the same explicit debug opt-in as the JavaScript
 * worklet. Production builds are silent by default, and callers must only pass
 * structural metadata: never payloads, identifiers, addresses, filesystem
 * paths, argv, keys, or exception messages.
 */
internal object ZMLog {
    private val isEnabled: Boolean
        get() = BuildConfig.ZAPP_MESSAGING_LOG_LEVEL.equals("debug", ignoreCase = true)

    fun debug(tag: String, message: () -> String) {
        if (isEnabled) Log.d(tag, message())
    }

    fun warning(tag: String, message: () -> String) {
        if (isEnabled) Log.w(tag, message())
    }

    fun error(tag: String, message: () -> String) {
        if (isEnabled) Log.e(tag, message())
    }
}

File: ios/Sources/ZappMessaging/Core/ZMLog.swift
//
//  ZMLog.swift
//  ZappMessaging
//

import Foundation

/// Privacy-safe native diagnostics shared by the Swift wrapper.
///
/// Logging is disabled unless the host explicitly selects the `debug` worklet
/// log level. Callers must log structural metadata only: never payloads,
/// identifiers, addresses, filesystem paths, argv, keys, or error descriptions.
enum ZMLog {
    private static let lock = NSLock()
    private static var enabled = false

    static func configure(level: String?) {
        lock.withLock {
            enabled = level?.lowercased() == "debug"
        }
    }

    static func debug(_ component: String, _ message: @autoclosure () -> String) {
        write(component, message: message)
    }

    static func warning(_ component: String, _ message: @autoclosure () -> String) {
        write(component, message: message)
    }

    static func error(_ component: String, _ message: @autoclosure () -> String) {
        write(component, message: message)
    }

    private static func write(_ component: String, message: () -> String) {
        let shouldWrite = lock.withLock { enabled }
        guard shouldWrite else { return }
        print("[ZappMessaging][\(component)] \(message())")
    }
}
